distribution/tools/entitlement-runtime/src/main/java/org/elasticsearch/entitlement/runtime/checks/EntitlementChecksService.java

```diff
@@ -0,0 +1,121 @@
+ */
+
+package org.elasticsearch.entitlement.runtime.checks;
+
+import org.elasticsearch.entitlement.checks.EntitlementChecks;
+import org.elasticsearch.entitlement.checks.EntitlementProvider;
+import org.elasticsearch.entitlement.runtime.api.Entitlement;
+import org.elasticsearch.entitlement.runtime.api.FlagEntitlement;
+import org.elasticsearch.entitlement.runtime.api.NotEntitledException;
+
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static java.util.Collections.newSetFromMap;
+import static org.elasticsearch.entitlement.runtime.internals.EntitlementInternals.isActive;
+import static org.elasticsearch.entitlement.runtime.internals.EntitlementInternals.isFrozen;
+
+/**
+ * Implementation of the {@link EntitlementChecks} interface.
+ * The trampoline module loads this object via SPI.
+ */
+public class EntitlementChecksService implements EntitlementChecks {
+    /**
+     * @return the same instance of {@link EntitlementChecksService} returned by {@link EntitlementProvider}.
+     */
+    public static EntitlementChecksService get() {
+        return (EntitlementChecksService) EntitlementProvider.checks();
+    }
+
+    private final Set<Module> entitledToExit = newSetFromMap(new ConcurrentHashMap<>());
+
+    public void grant(Module module, Entitlement e) {
+        if (isFrozen) {
+            throw new IllegalStateException("Entitlement grants are frozen");
+        }
+        switch (e) {
+            case FlagEntitlement f -> {
+                switch (f) {
+                    case EXIT_JVM -> entitledToExit.add(module);
+                }
+            }
+        }
+    }
+
+    /**
+     * Mainly for testing purposes.
+     */
+    public void revokeAll() {
+        if (isFrozen) {
+            throw new IllegalStateException("Entitlement grants are frozen");
+        }
+        entitledToExit.clear();
+    }
+
+    /**
+     * Causes entitlements to be enforced.
+     */
+    public void activate() {
+        isActive = true;
+    }
+
+    /**
+     * Disallows changes to the entitlement grants.
+     */
+    public void freeze() {
+        isFrozen = true;
+    }
+
+    @Override
+    public void checkSystemExit(Class<?> callerClass, System system, int status) {
+        checkFlagEntitlement(callerClass, entitledToExit);
+    }
+
+    private void checkFlagEntitlement(Class<?> callerClass, Set<Module> entitledModules) {
+        var requestingModule = requestingModule(callerClass);
+        if (isTriviallyAllowed(requestingModule)) {
+            // System.out.println(" - Trivially allowed");
+            return;
+        }
+        if (entitledModules.contains(requestingModule)) {
+            return;
+        }
+        throw new NotEntitledException("Missing entitlement for " + requestingModule);
+    }
+
+    private static Module requestingModule(Class<?> callerClass) {
+        if (callerClass != null) {
```

Comment:
when would the caller class be null? Shouldn't that be handled by instrumentation since that is what passes in the callerClass?

Reply:
I was anticipating situations where we want to instrument classes to which we can't add `@CallerSensitive`. In those cases, it might be that requiring `callerClass` to be non-null could mean the instrumentation would need to do another stack walk, and we might opt out of that optimization and just let the code do the normal stack walk it does anyway.